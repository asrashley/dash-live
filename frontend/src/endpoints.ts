import { createContext } from 'preact';
import log from 'loglevel';

import { routeMap, uiRouteMap} from '@dashlive/routemap';
import { CsrfTokenStore } from './CsrfTokenStore';
import { CsrfTokenCollection } from './types/CsrfTokenCollection';
import { AllManifests } from './types/AllManifests';
import { AllStreamsJson, AllStreamsResponse } from './types/AllStreams';
import { JWToken } from './types/JWToken';
import { MultiPeriodStream, MultiPeriodStreamJson } from './types/MultiPeriodStream';
import { DecoratedMultiPeriodStream } from "./types/DecoratedMultiPeriodStream";
import { MultiPeriodStreamSummary } from './types/MultiPeriodStreamSummary';
import { ContentRolesMap } from './types/ContentRolesMap';
import { ModifyMultiPeriodStreamJson, ModifyMultiPeriodStreamResponse } from './types/ModifyMultiPeriodStreamResponse';
import { InitialApiTokens } from './types/InitialApiTokens';
import { LoginRequest } from './types/LoginRequest';
import { LoginResponse } from './types/LoginResponse';
import { MultiPeriodStreamValidationRequest, MultiPeriodStreamValidationResponse } from './types/MpsValidation';
import { CgiOptionDescription } from './types/CgiOptionDescription';

type TokenStoreCollection = {
  files: CsrfTokenStore;
  kids: CsrfTokenStore;
  login: CsrfTokenStore;
  streams: CsrfTokenStore;
  upload: CsrfTokenStore;
};

export type ApiRequestOptions = {
  authorization?: string;
  body: RequestInit['body'];
  service: keyof TokenStoreCollection;
  signal?: AbortSignal;
  method: RequestInit['method'];
  rejectOnError?: boolean;
  query?: URLSearchParams;
};

export type GetAllStreamsProps = ApiRequestOptions & {
  withDetails: boolean
};

type RefreshAccessTokenResponse = {
  accessToken?: JWToken;
  csrfTokens: CsrfTokenCollection;
  ok: boolean;
  status: number;
};

export interface ApiRequestsProps extends Readonly<InitialApiTokens> {
  navigate: (url: string) => void;
}

export class ApiRequests {
  private csrfTokens: TokenStoreCollection;
  private accessToken: JWToken | null;
  private refreshToken: JWToken | null;
  private navigate: ApiRequestsProps['navigate'];

  constructor({csrfTokens, accessToken, refreshToken, navigate}: ApiRequestsProps) {
    this.csrfTokens = {
      files: new CsrfTokenStore(csrfTokens?.files),
      kids: new CsrfTokenStore(csrfTokens?.kids),
      login: new CsrfTokenStore(csrfTokens?.login),
      streams: new CsrfTokenStore(csrfTokens?.streams),
      upload: new CsrfTokenStore(csrfTokens?.upload),
    }
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.navigate = navigate;
  }

  getAllManifests(options: Partial<ApiRequestOptions> = {}): Promise<AllManifests> {
    return this.sendApiRequest<AllManifests>(routeMap.listManifests.url(), options);
  }

  getContentRoles(options: Partial<ApiRequestOptions> = {}): Promise<ContentRolesMap> {
    return this.sendApiRequest(routeMap.contentRoles.url(), options);
  }

  getCgiOptions(options: Partial<ApiRequestOptions> = {}) : Promise<CgiOptionDescription[]> {
    return this.sendApiRequest(routeMap.cgiOptions.url(), options);
  }

  async getUserInfo(signal: AbortSignal): Promise<LoginResponse | Response> {
    if (!this.refreshToken) {
      throw new Error('Cannot get user info without a refresh token');
    }
    const options: Partial<ApiRequestOptions> = {
      authorization: this.refreshToken.jwt,
      method: 'GET',
      rejectOnError: false,
      signal,
    };
    const response = await this.sendApiRequest<LoginResponse | Response>(routeMap.login.url(), options);
    if (response['success']) {
      this.accessToken = (response as LoginResponse).accessToken;
    }
    return response;
  }

  async loginUser(request: LoginRequest, options: Partial<GetAllStreamsProps> = {}): Promise<LoginResponse> {
    const response: LoginResponse = await this.sendApiRequest<LoginResponse>(routeMap.login.url(), {
      body: JSON.stringify(request),
      method: 'POST',
      service: 'login',
      ...options,
    });
    if (response.success) {
      if (response.accessToken) {
        this.accessToken = response.accessToken;
      }
      if (response.refreshToken) {
        this.refreshToken = response.refreshToken;
      }
    }
    return response;
  }

  async logoutUser(options: Partial<GetAllStreamsProps> = {}): Promise<Response> {
    const response: Response = await this.sendApiRequest(routeMap.login.url(), {
      method: 'DELETE',
      service: 'login',
      ...options,
    });
    this.accessToken = null;
    this.refreshToken = null;
    return response;
  }

  async getAllStreams(options: Partial<GetAllStreamsProps> = {}): Promise<AllStreamsResponse> {
    const { streams, keys } = await this.sendApiRequest<AllStreamsJson>(routeMap.listStreams.url(), options);
    return { streams, keys};
  }

  getAllMultiPeriodStreams(options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamSummary[]> {
    return this.sendApiRequest<MultiPeriodStreamSummary[]>(routeMap.listMps.url(), options);
  }

  async getMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStream> {
    const { model } = await this.sendApiRequest<MultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), options);
    return model;
  }

  async addMultiPeriodStream(
      data: DecoratedMultiPeriodStream,
      options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    const { errors, success, model } = await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.addMps.url(), {
      ...options,
      service,
      body: JSON.stringify({
        ...data,
        csrf_token
      }),
      method: 'PUT',
    });
    return { errors, success, model };
  }

  async modifyMultiPeriodStream(mps_name: string, data: DecoratedMultiPeriodStream,
     options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamJson> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      body: JSON.stringify({...data, csrf_token}),
      method: 'POST',
    });
  }

  async deleteMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<Response> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens
    );
    const query = new URLSearchParams({
      csrf_token,
    });
    return this.sendApiRequest<Response>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      query,
      method: 'DELETE',
    });
  }

  async validateMultiPeriodStream(data: MultiPeriodStreamValidationRequest,
        options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamValidationResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return this.sendApiRequest<MultiPeriodStreamValidationResponse>(routeMap.validateMps.url(), {
      ...options,
      method: 'POST',
      service: 'streams',
      body: JSON.stringify({...data, csrf_token}),
    });
  }

  private async sendApiRequest<T>(url: string, options: Partial<ApiRequestOptions>): Promise<T> {
    const { authorization, body, service, signal, method='GET',
      rejectOnError = true } = options;
    let { query } = options;
    let usedAccessToken = false;
    const headers: Headers = new Headers({
      "Content-Type": "application/json",
    });
    if (authorization !== undefined) {
      headers.set('Authorization', `Bearer ${authorization}`);
    } else {
      if (!this.accessToken && this.refreshToken) {
        await this.getAccessToken(signal);
      }
      if (this.accessToken) {
        headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
        usedAccessToken = true;
      } else if (service) {
        const token = await this.csrfTokens[service].getToken(signal);
        if (query === undefined) {
          query = new URLSearchParams({csrf_token: token});
        } else {
          query.set('csrf_token', token);
        }
      }
    }
    if (query) {
      url = `${url}?${query.toString()}`;
    }
    if (signal && signal.aborted) {
      throw signal.reason;
    }
    const cache = 'no-cache';
    const credentials = 'same-origin';
    const mode = 'same-origin';
    url = new URL(url, document.location.origin).href;
    log.trace(`sendApiRequest(${url})`);
    let fetchResult = await fetch(url, {
      cache,
      credentials,
      mode,
      body,
      headers,
      method,
      signal,
    });
    log.trace(`url=${url} status=${fetchResult.status} usedAccessToken=${usedAccessToken}`);
    if (fetchResult.status === 401 && usedAccessToken && this.refreshToken) {
      await this.getAccessToken(signal);
      headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
      log.trace(`retrying sendApiRequest(${url})`);
      fetchResult = await fetch(url, {
          cache,
          credentials,
          mode,
          body,
          headers,
          method,
          signal,
      });
    }
    if (!fetchResult.ok) {
      if (rejectOnError) {
        throw new Error(`${ url }: ${ fetchResult.status }`);
      } else {
        return fetchResult as T;
      }
    }
    if (fetchResult.status !== 200) {
      return fetchResult as T;
    }
    const data = await fetchResult.json();
    if (signal?.aborted) {
      throw signal.reason;
    }
    if (typeof(data?.csrfTokens) === "object") {
      this.updateCsrfTokens(data.csrfTokens);
    } else if(typeof(data?.csrf_tokens) === "object") {
      this.updateCsrfTokens(data.csrf_tokens);
    } else if (typeof(data?.csrf_token) === "string") {
      this.updateCsrfTokens({[service]: data.csrf_token});
    }

    return data as T;
  }

  private updateCsrfTokens(tokens: Partial<CsrfTokenCollection>) {
    for (const [key, value] of Object.entries(tokens)) {
      if (this.csrfTokens[key] === undefined) {
        this.csrfTokens[key] = new CsrfTokenStore(value);
      } else {
        this.csrfTokens[key].setToken(value);
      }
    }
  }

  private async getAccessToken(signal: AbortSignal): Promise<JWToken> {
    if (!this.refreshToken) {
      throw new Error('Cannot request an access token without a refresh token');
    }
    const options = {
      authorization: this.refreshToken.jwt,
      rejectOnError: false,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshAccessToken.url(), options);
    const { accessToken, csrfTokens, ok, status } = data ?? {};
    if (ok === false && status === 401) {
      this.navigate(uiRouteMap.login.url());
    }
    if (!accessToken) {
      throw new Error('Failed to refresh access token');
    }
    this.accessToken = accessToken;
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
    return accessToken;
  }

  private getCsrfTokens = async (signal: AbortSignal): Promise<void> => {
    if (!this.accessToken && this.refreshToken) {
      await this.getAccessToken(signal);
    }
    if (!this.accessToken) {
      throw new Error('Cannot request CSRF tokens without an access token');
    }
    const options = {
      authorization: this.accessToken.jwt,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshCsrfTokens.url(), options);
    const { csrfTokens } = data ?? {};
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
  };
}

export const EndpointContext = createContext<ApiRequests>(null);
