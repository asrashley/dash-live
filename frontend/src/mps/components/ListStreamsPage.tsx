import { type ComponentChildren } from 'preact';
import { Link } from 'wouter-preact';
import { useCallback, useContext } from 'preact/hooks'

import { useAllMultiPeriodStreams, AllMultiPeriodStreamsContext } from '../../hooks/useAllMultiPeriodStreams';
import { uiRouteMap } from '@dashlive/routemap';
import { SortIcon } from './SortIcon';
import { WhoAmIContext } from '../../user/hooks/useWhoAmI';

interface TableRowProps {
  name: string;
  title: string;
  periods: number[],
  duration: string;
}
function TableRow({name, title, periods, duration}: TableRowProps) {
  const url = uiRouteMap.editMps.url({mps_name: name});
  return <tr>
      <td className="name text-center">
        <Link to={ url }>{ name }</Link>
      </td>
      <td className="title text-center">
        <Link to={ url }>{ title }</Link>
      </td>
      <td className="num-periods text-end">
        { periods.length }
      </td>
      <td className="duration text-end">
        { duration }
      </td>
    </tr>;
}

interface SortableHeadingProps {
  name: string;
  children: ComponentChildren;
  className?: string;
}

function SortableHeading({name, children, className=""}: SortableHeadingProps) {
  const { sort, sortField, sortAscending } = useContext(AllMultiPeriodStreamsContext);
  const bold = name === sortField ? 'fw-bolder': '';
  const setSort = useCallback((ev: Event) => {
    ev.preventDefault();
    const ascending = name === sortField ? !sortAscending : true;
    sort(name, ascending);
  }, [name, sort, sortAscending, sortField]);

  return <th className={`${name} ${className}`}>
  <a onClick={setSort} className={`link-light ${bold}`}><SortIcon name={name} />{children}</a>
  </th>;
}

export default function ListStreamsPage() {
  const { user } = useContext(WhoAmIContext);
  const allMpsContext = useAllMultiPeriodStreams();
  const canModify = user.value.permissions.media === true;

  return <AllMultiPeriodStreamsContext.Provider value={allMpsContext}>
  <h1>Available Multi-period DASH streams</h1>
  <table className="table table-striped" id="mp-streams">
    <caption>Multi-Period Streams</caption>
    <thead>
      <tr>
        <SortableHeading name="name">Name</SortableHeading>
        <SortableHeading name="title">Title</SortableHeading>
        <th className="num-periods text-end">Periods</th>
        <th className="duration text-end">Duration</th>
      </tr>
    </thead>
    <tbody>
      { allMpsContext.streams.value.map(item => <TableRow key={item.name} {...item} />)}
    </tbody>
  </table>
  <div class="btn-toolbar">
    {canModify && <Link className="btn btn-primary btn-sm m-2"
      href={uiRouteMap.addMps.url()} >Add a Stream</Link> }
  </div>
</AllMultiPeriodStreamsContext.Provider>;
}
