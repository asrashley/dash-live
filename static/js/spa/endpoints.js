import { createContext } from 'preact';
import { routeMap } from '/libs/routemap.js';

class CsrfTokenStore {
  constructor(token) {
    this.token = token;
    this.pending = [];
  }

  setToken(token) {
    if (this.pending.length) {
      const {resolve} = this.pending.shift();
      resolve(token)
    } else {
      this.token = token;
    }
  }

  async getToken(signal) {
    const { token } = this;
    if (token) {
      this.token = null;
      return token;
    }
    const {promise, resolve, reject} = Promise.withResolvers();
    if (signal) {
      signal.addEventListener('abort', () => {
        reject(signal.reason);
      });
    }
    this.pending.push({resolve, reject});
    promise.finally(() => {
      this.pending = this.pending.filter(p => p.resolve !== resolve);
    });
    return await promise;
  }
}

export class ApiRequests {
  constructor({csrfTokens, accessToken, refreshToken}) {
    this.csrfTokens = {
      files: new CsrfTokenStore(csrfTokens?.files),
      kids: new CsrfTokenStore(csrfTokens?.kids),
      streams: new CsrfTokenStore(csrfTokens?.streams),
      upload: new CsrfTokenStore(csrfTokens?.upload),
    }
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
  }

  getAllStreams({ withDetails = false, ...options} = {}) {
    const query = withDetails ? new URLSearchParams({details:1}) : undefined;
    return this.sendApiRequest(routeMap.listStreams.url(), {
      service: 'streams',
      query,
      ...options,
    });
  }

  getAllMultiPeriodStreams(options) {
    return this.sendApiRequest(routeMap.listMps.url(), {
      service: 'streams',
      ...options,
    });
  }

  getMultiPeriodStream(mps_name, options) {
    return this.sendApiRequest(routeMap.editMps.url({mps_name}), {
      service: 'streams',
      ...options,
    });
  }

  addMultiPeriodStream(data, options) {
    return this.sendApiRequest(routeMap.addMps.url(), {
      ...options,
      service: 'streams',
      body: JSON.stringify(data),
      method: 'PUT',
    });
  }

  modifyMultiPeriodStream(mps_name, data, options) {
    return this.sendApiRequest(routeMap.editMps.url({mps_name}), {
      ...options,
      service: 'streams',
      body: JSON.stringify(data),
      method: 'POST',
    });
  }

  async sendApiRequest(url, options) {
    const { authorization, body, service, signal, method='GET' } = options;
    let { query } = options;
    if (this.csrfTokens[service] === undefined) {
      throw new Error(`Unknown service "${service}"`);
    }
    const headers = {
      "Content-Type": "application/json",
    }
    if (authorization !== undefined) {
      headers.Authorization = `Bearer ${authorization}`;
    } else {
      if (!this.accessToken && this.refreshToken) {
        await this.getAccessToken(signal);
      }
      if (this.accessToken) {
        headers.Authorization = `Bearer ${this.accessToken.jti}`;
      } else {
        const token = await this.csrfTokens[service].getToken();
        if (query === undefined) {
          query = new URLSearchParams({csrf_token: token});
        } else {
          query.set('csrf_token', token);
        }
      }
    }
    if (query) {
      url = `${url}?${query.toString()}`;
    }
    if (signal && signal.aborted) {
      throw signal.reason;
    }
    const fetchResult = await fetch(url, {
      cache: 'no-cache',
      credentials: 'same-origin',
      mode: 'same-origin',
      body,
      headers,
      method,
      signal,
    });
    if (!fetchResult.ok) {
      throw new Error(`Failed to fetch ${ url }: ${ fetchResult.status }`);
    }
    if (signal && signal.aborted) {
      throw signal.reason;
    }
    const data = await fetchResult.json();
    if (typeof(data?.csrfTokens) === "object") {
      this.updateCsrfTokens(data.csrfTokens);
    } else if(typeof(data?.csrf_tokens) === "object") {
      this.updateCsrfTokens(data.csrf_tokens);
    } else if (typeof(data?.csrf_token) === "string") {
      this.updateCsrfTokens({[service]: data.csrf_token});
    }

    return data;
  }

  updateCsrfTokens(tokens) {
    for (const [key, value] of Object.entries(tokens)) {
      if (this.csrfTokens[key] === undefined) {
        this.csrfTokens[key] = new CsrfTokenStore(value);
      } else {
        this.csrfTokens[key].setToken(value);
      }
    }
  }

  async getAccessToken(signal) {
    if (this.refreshToken === undefined) {
      throw new Error('Cannot request an access token without a refresh token');
    }
    const options = {
      authorization: this.refreshToken.jti,
      signal,
    };
    const data = await this.sendApiRequest(routeMap.getAccessToken, {}, options);
    const { accessToken } = data ?? {};
    if (accessToken) {
      this.accessToken = accessToken;
    }
  }
};

export function routeFromUrl(url) {
  for (const [name, route] of Object.entries(routeMap)) {
    if (route.re.test(url)) {
      return {
        name,
        ...route,
      };
    }
  }
  return undefined;
}

export const EndpointContext = createContext(null);
