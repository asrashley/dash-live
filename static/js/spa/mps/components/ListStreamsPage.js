import { html } from 'htm/preact';
import { Link } from 'wouter-preact';
import { useCallback, useContext } from 'preact/hooks'

import { Icon} from '@dashlive/ui';
import { useAllMultiPeriodStreams, AllMultiPeriodStreamsContext } from '@dashlive/hooks';
import { AppStateContext } from '../../appState.js';
import { routeMap } from '/libs/routemap.js';

function TableRow({name, title, periods, duration}) {
  const url = routeMap.editMps.url({mps_name: name});
  return html`
    <tr>
      <td class="name text-center">
        <${Link} to="${ url }">${ name }</${Link}>
      </td>
      <td class="title text-center">
        <${Link} to="${ url }">${ title }</${Link}>
      </td>
      <td class="num-periods text-end">
        ${ periods.length }
      </td>
      <td class="duration text-end">
        ${ duration }
      </td>
    </tr>
`;
}

function SortIcon({name}) {
  const { sortField, sortAscending } = useContext(AllMultiPeriodStreamsContext);
  const opacity = sortField === name ? 100 : 50;
  const iconName = (sortField === name && !sortAscending) ? "sort-alpha-up" : "sort-alpha-down";

  return html`<${Icon} name="${iconName}" className="opacity-${opacity}"/>`;
}

function SortableHeading({name, children, className=""}) {
  const { sort, sortField, sortAscending } = useContext(AllMultiPeriodStreamsContext);
  const bold = name === sortField ? 'fw-bolder': '';
  const setSort = useCallback((ev) => {
    ev.preventDefault();
    const ascending = name === sortField ? !sortAscending : true;
    sort(name, ascending);
  }, [name, sort, sortAscending, sortField]);

  return html`<th class="${name} ${className}">
  <a onClick=${setSort} class="link-light ${bold}"><${SortIcon} name=${name}/>
  ${children}</a>
  </th>`;
}

export function ListStreamsPage() {
  const { user } = useContext(AppStateContext);
  const allMpsContext = useAllMultiPeriodStreams();
  const canModify = user.value.permissions.media === true;

  return html`
<${AllMultiPeriodStreamsContext.Provider} value=${allMpsContext}>
  <h1>Available Multi-period DASH streams</h1>
  <table class="table table-striped" id="mp-streams">
    <caption>Multi-Period Streams</caption>
    <thead>
      <tr>
        <${SortableHeading} name="name">Name<//>
        <${SortableHeading} name="title">Title<//>
        <th className="num-periods text-end">Periods</th>
        <th className="duration text-end">Duration</th>
      </tr>
    </thead>
    <tbody>
      ${ allMpsContext.streams.value.map(item => html`<${TableRow} ...${item} />`)}
    </tbody>
  </table>
  <div class="btn-toolbar">
    ${canModify && html`<${Link} class="btn btn-primary btn-sm m-2"
      href=${routeMap.addMps.url()} >Add a Stream</${Link}>` }
  </div>
</${AllMultiPeriodStreamsContext.Provider}>`;
}
