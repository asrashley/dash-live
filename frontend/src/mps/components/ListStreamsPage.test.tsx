import { beforeEach, describe, expect, test } from "vitest";
import { mock } from "vitest-mock-extended";

import { renderWithProviders } from "../../test/renderWithProviders";
import ListStreamsPage from "./ListStreamsPage";
import { ApiRequests, EndpointContext } from "../../endpoints";

describe("ListStreamsPage", () => {
  const apiRequests = mock<ApiRequests>();
  const userInfo = {
    isAuthenticated: true,
    groups: ["MEDIA"],
  };

  beforeEach(() => {
    apiRequests.getAllMultiPeriodStreams.mockImplementation(async () => {
      const streams = await import("../../test/fixtures/multi-period-streams/index.json");
      return streams.default;
    });
  });

  test("should display list of multi-period streams", async () => {
    const { asFragment, findByText } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}><ListStreamsPage /></EndpointContext.Provider>,
      { userInfo }
    );
    await findByText("first title");
    await findByText("Add a Stream");
    expect(asFragment()).toMatchSnapshot();
  });

  test("should not show add stream button when user not logged in", async () => {
    const { queryByText, findByText } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}><ListStreamsPage /></EndpointContext.Provider>,
    );
    await findByText("first title");
    expect(queryByText("Add a Stream")).toBeNull();
  });
});
