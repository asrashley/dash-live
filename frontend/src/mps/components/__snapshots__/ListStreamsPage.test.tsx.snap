// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListStreamsPage > should display list of multi-period streams 1`] = `
<DocumentFragment>
  <h1>
    Available Multi-period DASH streams
  </h1>
  <table
    class="table table-striped"
    id="mp-streams"
  >
    <caption>
      Multi-Period Streams
    </caption>
    <thead>
      <tr>
        <th
          class="name "
        >
          <a
            class="link-light fw-bolder"
          >
            <span
              aria-label="sort-alpha-down icon"
              class="bi icon bi-sort-alpha-down opacity-100"
              role="img"
            />
            Name
          </a>
        </th>
        <th
          class="title "
        >
          <a
            class="link-light "
          >
            <span
              aria-label="sort-alpha-down icon"
              class="bi icon bi-sort-alpha-down opacity-50"
              role="img"
            />
            Title
          </a>
        </th>
        <th
          class="num-periods text-end"
        >
          Periods
        </th>
        <th
          class="duration text-end"
        >
          Duration
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="name text-center"
        >
          <a
            href="/multi-period-streams/demo"
          >
            demo
          </a>
        </td>
        <td
          class="title text-center"
        >
          <a
            href="/multi-period-streams/demo"
          >
            first title
          </a>
        </td>
        <td
          class="num-periods text-end"
        >
          2
        </td>
        <td
          class="duration text-end"
        >
          PT1M55S
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="btn-toolbar"
  >
    <a
      class="btn btn-primary btn-sm m-2"
      href="/multi-period-streams/.add"
    >
      Add a Stream
    </a>
  </div>
</DocumentFragment>
`;
